### PR TITLE
Track MET Emulation

### PR DESCRIPTION
**Pull request by Chriisbrown**
#### PR description:
This PR adds functionality for the Track MET Emulation using HLS integer types throughout.
The MET emulation itself is contained within the plugin L1TrackerEtMissEmulatorProducer.cc that makes use of a set of functions and constants defined in L1TkEtMissEmuAlgo.h

The cordic sqrt used for integer square root calculation has a separate class as does the input track conversion.

This PR relies on https://github.com/cms-l1t-offline/cmssw/pull/896 as it makes use of the new Trackword. It also makes has the option of making use of the GTTinputproducers from the same PR.

Additional modifications are needed to make use of the vertex emulator from here https://github.com/cms-l1t-offline/cmssw/pull/901 however the current producer works without the need for this by using the floating point vertex producer.

There is an additional change made to the TkEtMiss dataformat to include two new variables for validation use: TrkMET Phi and Quality, also included is a new constructor to facilitate the use of these variables. This has also meant a small update to the L1TrackerEtMissProducer.cc 

Finally there is an EDAnalyser for verify the output of the Emulation

#### PR validation:

Checked versus floating point Track MET in both MET and MET phi residuals as well as rates and efficiencies with high levels of agreement.

Plots shown here: https://indico.cern.ch/event/1021423/

Code & style checks have been performed as per https://twiki.cern.ch/twiki/bin/view/CMS/L1TrackCodeRules

To run use test/L1TrackMET_cfg.py

### CODE REVIEW

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+#include <filesystem>
+#include <fstream>
+#include <ap_int.h>
+
+using namespace std;
+// Namespace that defines constants and types used by the EtMiss Emulation
+// Includes functions for writing LUTs and converting to integer representations
+namespace L1TkEtMissEmuAlgo {
+
+  const unsigned int N_chi2XYbits_{4};  //Taken from TTTrack_Word class
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: Why not take all of these constants and typedefs from the TTTrack_TrackWord class?

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  typedef ap_uint<N_z0Bits_> iZ0;
+  typedef ap_uint<N_ptBits_> iPt;
+  typedef ap_int<N_ptBits_ + N_EtExtra_> iEt;
+  typedef ap_uint<N_etaBits_> iEta;
+
+  typedef ap_uint<N_METbits_> iMET;
+  typedef ap_uint<N_METphibits_>
+      iMETphi;  //Cordic means this is evaluated between 0 and 2Pi rather than -pi to pi so unsigned
+
+  /*
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: You'll probably want to remove any code which is commented out.

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  const unsigned int N_globPhiBins_ = 1 << N_globphiBits_;
+  const unsigned int N_METbins_ = 1 << N_METbits_;
+  const unsigned int N_METphibins_ = 1 << N_METphibits_;
+
+  const unsigned int N_etaregions_{6};
+  const unsigned int N_sectors_{9};
+  const unsigned int N_quadrants_{4};  //4 quadrants + 1
+
+  const float max_TWord_Pt_{1024};         //21955               // Arbitrary, define properly!
+  const float max_TWord_Phi_{0.69813248};  // relative to the center of the sector
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: You can also take these next three lines and the chi2 bins from the TTTrack_TrackWord class.

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+public:
+  L1TkEtMissEmuTrackTransform() = default;  
+  ~L1TkEtMissEmuTrackTransform() {};
+
+  void GenerateLUTs();  //Generate internal LUTs needed for track transfrom
+
+  InternalEtWord TransformTrack(TTTrack< Ref_Phase2TrackerDigi_ >& track_ref,float PV);
+  InternalEtWord TransfromCuts(const edm::ParameterSet& iConfig);
+
+    //Converts local int phi to global int phi
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h]: The spacing in front of some of the comments does not match the spacing in front of the code.

File: DataFormats/L1TCorrelator/src/classes_def.xml [hunk 5]
```diff
@@ -9,8 +9,8 @@
   <class name="edm::Ref<std::vector<l1t::TkPrimaryVertex>,l1t::TkPrimaryVertex,edm::refhelper::FindUsingAdvance<std::vector<l1t::TkPrimaryVertex>,l1t::TkPrimaryVertex> >"/>
   <class name="std::vector<edm::Ref<std::vector<l1t::TkPrimaryVertex>,l1t::TkPrimaryVertex,edm::refhelper::FindUsingAdvance<std::vector<l1t::TkPrimaryVertex>,l1t::TkPrimaryVertex> > >"/>
 
-  <class name="l1t::TkEtMiss" ClassVersion="3">
-   <version ClassVersion="3" checksum="77762919"/>
+  <class name="l1t::TkEtMiss" ClassVersion="5">
```
Comment (aperloff) [inline DataFormats/L1TCorrelator/src/classes_def.xml]: Why delete the `ClassVersion="3"`? Also, what happened to `ClassVersion="4"`?

File: L1Trigger/L1TTrackMatch/interface/METTrackTransform.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
+#ifndef L1Trigger_L1TTrackMatch_METTrackTransform_HH
+#define L1Trigger_L1TTrackMatch_METTrackTransform_HH
+
+#include "L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h"
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+using namespace L1TkEtMissEmuAlgo;
+
+struct InternalEtWord{
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/METTrackTransform.h]: Why is this `InternalEtWord` defined both in this file and in L1TkEtMisEmuTrackTransorm? If this is really the same struct, then you should define it only in one location and then include it. Also, if this is internal to some other class/struct/namespace, then you should define it within that namespace.

File: L1Trigger/L1TTrackMatch/interface/METTrackTransform.h [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  std::map<std::string, float> maxmap { {"pt",max_TWord_Pt_},
+                                        {"eta",max_TWord_Eta_},
+                                        {"z0",max_TWord_Z0_},
+                                        {"phi",max_TWord_Phi_}};
+
+
+  std::vector<iglobPhi> phi_quadrants;
+  std::vector<iglobPhi> phi_shift;
+
+  const float chi2Values[N_chi2XYbins_] = {0, 0.25, 0.5, 1, 2, 3, 5, 7, 10, 20, 40, 100, 200, 500, 1000, 3000};
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/interface/METTrackTransform.h]: Again, this seems to me to be a duplicate definition and should be taken from TTTrack_TrackWord.

File: L1Trigger/L1TTrackMatch/src/L1TkEtMissEmuTrackTransform.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
+#include <cmath>
+#include "L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h"
+
+void L1TkEtMissEmuTrackTransform::GenerateLUTs(){
+    phi_quadrants = generate_phi_slice_LUTs (L1TkEtMissEmuAlgo::N_quadrants_);
+    phi_shift     = generate_phi_slice_LUTs (L1TkEtMissEmuAlgo::N_sectors_);
+}
+
+InternalEtWord L1TkEtMissEmuTrackTransform::TransformTrack(TTTrack< Ref_Phase2TrackerDigi_ >& track_ref, float PV){
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/src/L1TkEtMissEmuTrackTransform.cc]: Seems to be redundant now that the TTTrack_TrackWord is "finished." Is this on purpose?

File: L1Trigger/L1TTrackMatch/src/METTrackTransform.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
+#include <cmath>
+#include "L1Trigger/L1TTrackMatch/interface/METTrackTransform.h"
+
+void METTrackTransform::GenerateLUTs(){
+    phi_quadrants = generate_phi_slice_LUTs (L1TkEtMissEmuAlgo::N_quadrants_);
+    phi_shift     = generate_phi_slice_LUTs (L1TkEtMissEmuAlgo::N_sectors_);
+}
+
+InternalEtWord METTrackTransform::TransformTrack(TTTrack< Ref_Phase2TrackerDigi_ >& track_ref, float PV){
```
Comment (aperloff) [inline L1Trigger/L1TTrackMatch/src/METTrackTransform.cc]: Same comment about possibly being redundant.

Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/src/METTrackTransform.cc]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/src/L1TkEtMissEmuTrackTransform.cc (hunk 8 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/src/L1TkEtMissEmuTrackTransform.cc]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/interface/METTrackTransform.h (hunk 7 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/METTrackTransform.h]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/interface/METTrackTransform.h (hunk 6 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/METTrackTransform.h]: Fixed in latest commit, only L1TkEtMissEmuTrackTransform remains

File: DataFormats/L1TCorrelator/src/classes_def.xml (hunk 5 above)
Comment (Chriisbrown) [reply DataFormats/L1TCorrelator/src/classes_def.xml]: Error on my part, this has been resolved

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h (hunk 4 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuTrackTransform.h]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h (hunk 3 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h (hunk 2 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: Fixed in latest commit

File: L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h (hunk 1 above)
Comment (Chriisbrown) [reply L1Trigger/L1TTrackMatch/interface/L1TkEtMissEmuAlgo.h]: Fixed in latest commit

**Comment by aperloff** [comment]
Thanks @Chriisbrown. I'm happy with the changes.

File: L1Trigger/L1TTrackMatch/python/L1TrackerEtMissProducer_cfi.py [hunk 10]
```diff
@@ -1,12 +1,15 @@
 import FWCore.ParameterSet.Config as cms
+from L1Trigger.VertexFinder.VertexProducer_cff import VertexProducer
 
 L1TrackerEtMiss = cms.EDProducer('L1TrackerEtMissProducer',
     L1TrackInputTag = cms.InputTag("TTTracksFromTrackletEmulation", "Level1TTTracks"),
-    L1VertexInputTag = cms.InputTag("L1TkPrimaryVertex"),
+    L1VertexInputTag = cms.InputTag("VertexProducer", VertexProducer.l1VertexCollectionName.value()),
```
Comment (gpetruc) [inline L1Trigger/L1TTrackMatch/python/L1TrackerEtMissProducer_cfi.py:6]: Hi.

This module assumes the PV producer is called `VertexProducer`, but the L1 sequence in [SimL1Emulator_cff.py](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-CMSSW_11_1_7/L1Trigger/Configuration/python/SimL1Emulator_cff.py#L119) currently has it as `L1VertexFinder`, and indeed when running the standard sequence I get a bunch of warnings of the sort
````
%MSG-e L1TrackerEtMissProducer:  L1TrackerEtMissProducer:L1TrackerEtMiss  17-May-2021 10:45:22 CEST Run: 1 Event: 186111

Warning: VertexCollection not found in the event. Exit
````

@rekovic @emacdonald16 @aperloff

Comment (aperloff) [reply L1Trigger/L1TTrackMatch/python/L1TrackerEtMissProducer_cfi.py:6]: @gpetruc The change should happen in https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-CMSSW_11_1_7/L1Trigger/Configuration/python/SimL1Emulator_cff.py. All you would need would be a line like:
```python
L1TrackEtMiss.L1VertexInputTag = cms.InputTag("L1VertexFinder", VertexProducer.l1VertexCollectionName.value())
```

**Pull request merged**